### PR TITLE
feat(gas-oracle & rollup-relayer): graceful shutdown

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

Graceful shutdown for `rollup-relayer` and `gas-oracle`. when receiving a shutdown signal, stop all loops, but wait for pending transactions to get confirmed before shutdown. (otherwise, a sudden restart from devs or Kubernetes might leave pending transactions in the DB).

The graceful shutdown is implemented by the following steps:

1. Stop monitoring new blocks and adding new transactions to senders.
2. Invoke `relayer.Stop()` and wait for `sender.pendingTxs` to be cleared.
3. Cease receiving confirmation signals from the sender.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

**Review by zimpha — DISMISSED** [review]

File: rollup/cmd/gas_oracle/app/app.go [hunk 1]
```diff
@@ -106,9 +97,31 @@ func action(ctx *cli.Context) error {
...
+
+		// Prevent new transactions by canceling the loop context.
+		loopCancel()
+
+		// Close relayers to ensure all pending transactions are processed.
+		// This includes any in-flight transactions that have not yet been confirmed.
+		l1relayer.Close()
+		l2relayer.Close()
+
+		// Halt confirmation signal handling by canceling the instance context.
```
Comment (Thegaram) [inline rollup/cmd/gas_oracle/app/app.go]: What do you mean by this comment?

File: rollup/cmd/gas_oracle/app/app.go [hunk 2]
```diff
@@ -106,9 +97,31 @@ func action(ctx *cli.Context) error {
...
+		instanceCancel()
+
+		// Close the database connection.
+		if err = database.CloseDB(db); err != nil {
+			log.Error("Failed to close database connection", "error", err)
+		}
+
+		log.Info("Graceful shutdown done")
```
Comment (Thegaram) [inline rollup/cmd/gas_oracle/app/app.go:125]: Important to node that `docker stop` and `docker restart` will kill the process if it doesn't terminate within a certain period. We'll also need to adjust this by passing a flag with a longer period.

**Review by Thegaram — DISMISSED** [review]

File: rollup/cmd/gas_oracle/app/app.go (hunk 1 above)
Comment (colinlyguo) [reply rollup/cmd/gas_oracle/app/app.go]: It's redundant right now, add it here for completeness' sake. it is intended to end handling confirmations in the following places:
https://github.com/scroll-tech/scroll/blob/v4.3.50/rollup/internal/controller/relayer/l1_relayer.go#L142
https://github.com/scroll-tech/scroll/blob/v4.3.50/rollup/internal/controller/relayer/l2_relayer.go#L697

File: rollup/cmd/gas_oracle/app/app.go (hunk 2 above)
Comment (colinlyguo) [reply rollup/cmd/gas_oracle/app/app.go:125]: oh, I see. so do we also need to change Jenkins scripts in testnet repo?

File: rollup/cmd/gas_oracle/app/app.go [hunk 3]
```diff
@@ -50,49 +50,40 @@ func action(ctx *cli.Context) error {
 	if err != nil {
 		log.Crit("failed to load config file", "config file", cfgFile, "error", err)
 	}
-	subCtx, cancel := context.WithCancel(ctx.Context)
+	instanceCtx, instanceCancel := context.WithCancel(ctx.Context)
+	loopCtx, loopCancel := context.WithCancel(ctx.Context)
```
Comment (georgehao) [inline rollup/cmd/gas_oracle/app/app.go:54]: why create two ctx ?

Comment (colinlyguo) [reply rollup/cmd/gas_oracle/app/app.go:54]: The `instanceCtx` is for managing the lifecycle of instances: `handleConfirmLoop` should be closed after `sender.Stop()`.

The `loopCtx` is for managing the lifecycle of polling (which could send new transactions): `xxxSender.SendTransaction` should be closed first.

**Review by zimpha — DISMISSED** [review]

File: rollup/cmd/gas_oracle/app/app.go (hunk 2 above)
Comment (Thegaram) [reply rollup/cmd/gas_oracle/app/app.go:125]: Yes, like this: https://github.com/scroll-tech/testnet/pull/840 (currently only configured for `l1geth`).

File: rollup/cmd/gas_oracle/app/app.go (hunk 1 above)
Comment (Thegaram) [reply rollup/cmd/gas_oracle/app/app.go]: I just find the phrasing "halt confirmation signal" confusing. I guess it means "Trigger relayer graceful shutdown".

File: rollup/cmd/gas_oracle/app/app.go (hunk 2 above)
Comment (colinlyguo) [reply rollup/cmd/gas_oracle/app/app.go:125]: OK. will raise a PR too for testing.

File: rollup/cmd/gas_oracle/app/app.go (hunk 1 above)
Comment (colinlyguo) [reply rollup/cmd/gas_oracle/app/app.go]: actually the graceful shutdown is triggered by:
```
		// Close relayers to ensure all pending transactions are processed.
		// This includes any in-flight transactions that have not yet been confirmed.
		l1relayer.Close()
		l2relayer.Close()
```
will think a better comment.

**Comment by colinlyguo** [comment]
@zimpha the contract CI failure is a bit weird. it shows several `*.sol` files' coverage has decreased.

File: rollup/cmd/gas_oracle/app/app.go (hunk 1 above)
Comment (colinlyguo) [reply rollup/cmd/gas_oracle/app/app.go]: rephrased the comment in [8d94bbe](https://github.com/scroll-tech/scroll/pull/1069/commits/8d94bbe43eef87b1f11f8d38764cb8bccd27f2de)

Comment (colinlyguo) [reply rollup/cmd/gas_oracle/app/app.go]: The graceful shutdown is implemented by the following steps:
1. Stop monitoring new blocks and adding new transactions to senders.
2. Invoke `relayer.Stop()` and wait for `sender.pendingTxs` to be cleared.
3. Cease receiving confirmation signals from the sender.

**Comment by georgehao** [comment]
I think this pr don't resolve the graceful shutdown completely. The send's data is stored in memory, when restart, need a lot of time to resume. But we can do it in another pr

**Comment by colinlyguo** [comment]
Replaced with the graceful restart feature, which entirely removes the `pendingTxs` `sync.Map`: https://github.com/scroll-tech/scroll/pull/1076

**Pull request closed without merging**